### PR TITLE
gh-119189:  Fix the fraction module so that __rpow__ works on arbitrary classes.

### PR DESCRIPTION
**Pull request by zitterbewegung**
# gh-119189: Fix the fraction module so that __rpow__ works on arbitrary classes.

```
 The fraction module implicitly casted float if __rpow__ was implemented in your class when raising a fraction to a power. Now we replace the cast to float and instead check if it is a float or complex  and if the class isn't it will raise an exception.
```

### CODE REVIEW

**Review by serhiy-storchaka — APPROVED** [review]
Please update tests.

Also, I think that the text of the NEWS entry should be rewritten. It's hard to understand right now, I'm not even sure if it's correct at all.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Please update tests and rewrite the NEWS entry.

**Comment by zitterbewegung** [comment]
I don't understand what should the tests be updated to?

**Comment by zitterbewegung** [comment]
I have made the requested changes; please review again.

**Comment by zitterbewegung** [comment]
I have made the requested changes; please review again.

File: Misc/NEWS.d/next/Library/2024-05-20-13-48-37.gh-issue-119189.dhJVs5.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+When using the ** operator with a Fraction as a base and an object that implements __rpow__ as an exponent the fraction it is now not cast to a float.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2024-05-20-13-48-37.gh-issue-119189.dhJVs5.rst]:
```suggestion
When using the ``**`` operator or :func:`pow` with :class:`~fractions.Fraction`
as the base and an exponent that is not rational, a float, or a complex, the
fraction is no longer converted to a float.
```

File: Lib/test/test_fractions.py [hunk 2]
```diff
@@ -1,5 +1,3 @@
-"""Tests for Lib/fractions.py."""
-
```
Comment (serhiy-storchaka) [inline Lib/test/test_fractions.py]: This is an unrelated change.

Comment (zitterbewegung) [reply Lib/test/test_fractions.py]: Fixed in [b585219](https://github.com/python/cpython/pull/119242/commits/b585219ac44c740baaec336ef6b07a17989f050a)

**Comment by zitterbewegung** [comment]
I have made the requested changes; please review again.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. Thank you for your contribution @zitterbewegung.

**Comment by serhiy-storchaka** [comment]
I consider this a bug fix.
